src/useMedia.ts

```diff
@@ -0,0 +1,38 @@
+import { useEffect, useState, useRef } from 'react';
+import { canUseDOM, mockMediaQueryList } from './utils';
+
+/**
+ * Tracks media query match state.
+ *
+ * @param query media query string for evaluation
+ *
+ * @returns {boolean} isMatch
+ *
+ * @example
+ * function Example() {
+ *   const isWidescreen = useMedia('(min-width: 400px)');
+ *   // ...
+ * }
+ */
+export default function useMedia(query: string): boolean {
+  const queryListRef = useRef<MediaQueryList>(
+    canUseDOM ? window.matchMedia(query) : mockMediaQueryList,
+  );
+  const [isMatch, setIsMatch] = useState<boolean>(
+    () => queryListRef.current.matches,
+  );
+
+  useEffect(() => {
+    const queryList = queryListRef.current;
+    const handler = () => {
+      setIsMatch(window.matchMedia(query).matches);
+    };
+
+    queryList.addListener(handler);
+    return () => {
+      queryList.removeListener(handler);
+    };
+  }, [query]);
+
+  return isMatch;
+}
```

Comment:
Here is a simpler implementation, which re-uses `managedEventListener`, as [made possible](https://drafts.csswg.org/cssom-view/#mediaquerylist-event-summary) by the CSSOM View Module spec:

```ts
import { useEffect, useState } from 'react';
import { canUseDOM, managedEventListener } from './utils';

export default function useMedia(query: string): boolean {
  const mediaQueryList = canUseDOM && matchMedia(query);
  const [matches, setMatches] = useState(
    mediaQueryList && mediaQueryList.matches,
  );

  useEffect(() =>
    mediaQueryList
      ? managedEventListener(mediaQueryList, 'change', ((
          event: MediaQueryListEvent,
        ) => {
          setMatches(event.matches);
        }) as EventListener)
      : undefined,
  );

  return matches;
}
```

Reply:
Only issue with this is that mediaQueryList will reevaluate every render unnecessarily. I've used a `ref` to resolve this in the most recent commit.

Reply:
Unless the contents of `mediaQueryList` are enclosed in a function, they still get evaluated on every rerender, even when using a ref. We may want to `useMemo` as a replacement, instead.